groth16_verifier/src/lib.rs

```diff
@@ -0,0 +1,52 @@
+#![no_std]
+use soroban_sdk::{
+    contract, contractimpl, contracttype,
+    crypto::bls12_381::{Fr, G1Affine, G2Affine},
+    vec, Env, Vec,
+};
+
+#[derive(Clone)]
+#[contracttype]
+pub struct VerificationKey {
+    pub alpha: G1Affine,
+    pub beta: G2Affine,
+    pub gamma: G2Affine,
+    pub delta: G2Affine,
+    pub ic: Vec<G1Affine>,
+}
+
+#[derive(Clone)]
+#[contracttype]
+pub struct Proof {
+    pub a: G1Affine,
+    pub b: G2Affine,
+    pub c: G1Affine,
+}
+
+#[contract]
+pub struct Groth16Verifier;
+
+#[contractimpl]
+impl Groth16Verifier {
+    pub fn verify_proof(env: Env, vk: VerificationKey, proof: Proof, pub_signals: Vec<Fr>) -> bool {
+        let bls = env.crypto().bls12_381();
+
+        // vk_x = IC0 + pub_signals[0]*IC1
+        // Here we have only one public input. If there were more, we'd sum them similarly.
+        let mut vk_x = vk.ic.get(0).unwrap();
+        if let Some(s) = pub_signals.get(0) {
```

Comment:
Does the API usage to get the first and ignore the rest present an API trap to developers copying this example?

Would it be safer for contracts like this if we instead converted the `vk.ic` `Vec` into an array of length `1`, or even better a one element tuple? That should, I think, ensure that the passed in `Vec` was length `1` (please double check that if you rely on it).

Reply:
Good point. This was an initial over-simplified version, fixed it to account for generic number of public inputs.